demo/Render.hpp

```diff
@@ -0,0 +1,841 @@
+ * @param[in,out] mesh holds vertex data and opengl handles
+ */
+void Allocate(Mesh& mesh);
+
+/**
+ * @brief Removes given mesh from the GPU memory and updates its handles
+ * @param[in,out] mesh holds vertex data and opengl handles
+ */
+void Deallocate(Mesh& mesh);
+
+/**
+ * @brief Makes mesh and copies its data to the GPU memory
+ * @param[in] vertices vertex data
+ * @param[in] indices index data
+ * @return created mesh
+ */
+Mesh Create(std::vector<GLdouble>&& vertices, std::vector<GLuint>&& indices);
+
+/**
+ * @brief Makes line segment mesh and copies its data to the GPU memory
+ * @param[in] start first point of the line segment
+ * @param[in] end second point of the line segment
+ * @return created mesh
+ */
+Mesh CreateLineSegment(glm::vec3 start, glm::vec3 end);
+
+/**
+ * @brief Makes plane mesh and copies its data to the GPU memory
+ * @param[in] normal plane normal
+ * @param[in] length length of the side of the plane
+ * @return created mesh
+ */
+Mesh CreatePlane(glm::dvec3 normal, double length);
+
+/**
+ * @brief Makes ICO sphere mesh and copies its data to the GPU memory
+ * @param[in] radius sphere's radius
+ * @param[in] depth sphere's level of the details
+ * @return created mesh
+ */
+Mesh CreateSphere(double radius, uint32_t depth);
+
+/**
+ * @brief Makes box mesh and copies its data to the GPU memory
+ * @param[in] i orthogonal basis vector of the box
+ * @param[in] j orthogonal basis vector of the box
+ * @param[in] k orthogonal basis vector of the box
+ * @return created mesh
+ */
+Mesh CreateBox(glm::dvec3 i, glm::dvec3 j, glm::dvec3 k);
+
+/**
+ * @brief Deallocates GPU and local memory used by the mesh
+ * @param[in] mesh reference to the mesh
+ */
+void Delete(Mesh& mesh);
+} // namespace mesh
+
+namespace asset
+{
+/**
+ * @brief Represents a unit of storage with id
+ * @tparam T storage data type
+ */
+template < typename T >
+struct Asset
+{
+    Handle id;
+    T data;
+};
+
+/**
+ * @brief Finds first available or creates new element for the given asset vector and returns its handle
+ * @tparam T storage data type
+ * @param[in] data reference to the vector of assets
+ * @return hande to the free element
+ */
+template < typename T >
+Handle Make(std::vector<Asset<T>>& data)
+{
+    for (size_t i = 0; i < data.size(); ++i)
+    {
+        if (data[i].id == 0)
+        {
+            data[i].id = static_cast<Handle>(i + 1);
+            return data[i].id;
+        }
+    }
+
+    data.resize(data.size() + 1);
+    data.back().id = static_cast<Handle>(data.size());
+    return data.back().id;
+}
+
+/**
+ * @brief Return reference to the data assigned to the given vector
+ *
+ * @attention Reference might be invalidated after the push to the given vector
+ *
+ * @tparam T storage data type
+ * @param[in] data assets vector
+ * @param[in] id handle of the data
+ * @return reference to the data
+ */
+template < typename T >
+T& Get(std::vector<Asset<T>>& data, Handle id)
```

Comment:
Why this stuff is not a part of a class? `std::vector<Asset<T>>` could've been a private field. Is this just some private stuff that shouldn't really be used in user code and could be stored in a separate file? If it's not intended to be used in user code, why is it in the top of the file?